### PR TITLE
Remove python bindings for `torch.slice`

### PR DESCRIPTION
**Pull request by sethah**
This patch is a pivot from https://github.com/pytorch/pytorch/pull/7902. `torch.slice` is not intended to be used directly since its functionality is encompassed within tensor indexing notation. We should remove the function from the Python API.

### CODE REVIEW

**Comment by sethah** [comment]
I am not super familiar with the code generation mechanics, so any help is appreciated if there is more to be done in order to fully remove this from the API.

**Comment by sethah** [comment]
cc @colesbury

File: test/test_torch.py [hunk 1]
```diff
@@ -3314,24 +3314,6 @@ def test_randn(self):
         torch.randn(SIZE, SIZE, out=res2)
         self.assertEqual(res1, res2)
 
-    def test_slice(self):
```
Comment (sethah) [inline test/test_torch.py]: This reduces test coverage. Would it make sense to simply leave the test, but convert  to index notation?

Comment (soumith) [reply test/test_torch.py]: yes, that makes a lot of sense to do

**Review by soumith — APPROVED** [review]

**Comment by vadimkantorov** [comment]
Should this be reverted?

There's a usecase of doing a slice (supporting step/stride) on a dynamic dim, described here: https://discuss.pytorch.org/t/use-python-like-slice-indexing-across-a-given-dimension/89606/8. I have to resort to some hacks like:
```python
def tensor_dim_slice(tensor, dim, s):
    return tensor[(slice(None),) * (dim if dim >= 0 else dim + tensor.dim()) + (s, )]
```

This is needed for https://github.com/pytorch/pytorch/issues/32867

Otherwise, changing `narrow` for supporting step/stride could be a solution as well

@albanD

**Comment by albanD** [comment]
I think narrow is the right API to do this. Slice would just be doing the same thing with a slightly different set of arguments.

**Comment by vadimkantorov** [comment]
What's good for slice is flexible support of None (for absent start/stop). I think this is missing from narrow, since length must be computed - this is not a big deal, but maybe for tracing it's nasty?

**Comment by t-vi** [comment]
I don't think making narrow (which essentially is something we got from Torch7) into a not-so-nice slice is good for consistency between C++ / TorchScript (`.code`), and Python.

**Pull request merged**